diff --git a/Makefile.ctr b/Makefile.ctr
index b182ecf..382781a 100644
--- a/Makefile.ctr
+++ b/Makefile.ctr
@@ -27,6 +27,12 @@ APP_SYSTEM_MODE_EXT  = 124MB
 APP_BIG_TEXT_SECTION = 0
 APP_USE_SVCHAX       = 0
 
+ifeq ($(VCOVEN_ARCADE_ROMFS),1)
+APP_ROMFS            ?= $(CURDIR)/vcoven-romfs
+CFLAGS               += -DVCOVEN_ARCADE_ROMFS
+CFLAGS               += -DUSE_CTRULIB_2
+endif
+
 include pkg/ctr/Makefile.cores
 
 OBJ :=
@@ -154,6 +160,10 @@ MAKEROM_ARGS_COMMON = -rsf $(APP_RSF) -exefslogo -elf $(TARGET).elf -icon $(TARG
 	-DAPP_SYSTEM_MODE=$(APP_SYSTEM_MODE) -DAPP_SYSTEM_MODE_EXT=$(APP_SYSTEM_MODE_EXT) \
 	-major "$(APP_VERSION_MAJOR)" -minor "$(APP_VERSION_MINOR)" -micro "$(APP_VERSION_MICRO)"
 
+ifneq ($(strip $(APP_ROMFS)),)
+MAKEROM_ARGS_COMMON += -DAPP_ROMFS=$(APP_ROMFS)
+endif
+
 INCDIRS := -I$(CTRULIB)/include
 LIBDIRS := -L. -L$(CTRULIB)/lib
 
diff --git a/ctr/ctr_system.c b/ctr/ctr_system.c
index e310489..6b8774b 100644
--- a/ctr/ctr_system.c
+++ b/ctr/ctr_system.c
@@ -5,6 +5,10 @@
 #include <unistd.h>
 #include <string.h>
 #include <stdlib.h>
+#ifdef VCOVEN_ARCADE_ROMFS
+#include <3ds/romfs.h>
+#include <dirent.h>
+#endif
 #include "ctr_debug.h"
 
 #define CTR_APPMEMALLOC_PTR ((u32*)0x1FF80040)
@@ -235,9 +239,32 @@ void __system_initArgv(void)
    }
    else
    {
+#ifdef VCOVEN_ARCADE_ROMFS
+      DIR *dir = opendir("romfs:/content");
+      struct dirent *entry;
+      char *content = NULL;
+      while (dir && (entry = readdir(dir)))
+      {
+         if (!strcmp(entry->d_name, ".") || !strcmp(entry->d_name, ".."))
+            continue;
+         size_t length = strlen(entry->d_name) + sizeof("romfs:/content/");
+         content = (char*)malloc(length);
+         snprintf(content, length, "romfs:/content/%s", entry->d_name);
+         break;
+      }
+      if (dir)
+         closedir(dir);
+      if (!content)
+         content = "romfs:/content/game.zip";
+      __system_argc    = 2;
+      __system_argv    = (char**) malloc(sizeof(char**) * 3);
+      __system_argv[0] = "sdmc:/retroarch/retroarch";
+      __system_argv[1] = content;
+#else
       __system_argc    = 1;
       __system_argv    = (char**) malloc(sizeof(char**) * 2);
       __system_argv[0] = "sdmc:/retroarch/retroarch";
+#endif
    }
    __system_argv[__system_argc] = NULL;
 }
@@ -249,6 +276,9 @@ void initSystem(void (*retAddr)(void))
 
    __libctru_init(retAddr);
    __appInit();
+#ifdef VCOVEN_ARCADE_ROMFS
+   romfsInit();
+#endif
    __system_initArgv();
    __libc_init_array();
 }
@@ -256,6 +286,9 @@ void initSystem(void (*retAddr)(void))
 void __attribute__((noreturn)) __ctru_exit(int rc)
 {
    __libc_fini_array();
+#ifdef VCOVEN_ARCADE_ROMFS
+   romfsExit();
+#endif
    __appExit();
    asm ("mov sp, %[saved_stack] \n\t" : : [saved_stack] "r" (__saved_stack));
    __libctru_exit(rc);
